### DISCUSSION TITLE
Feature request: Support OpenAI's MacOS ChatGPT desktop "Work With" feature

### DISCUSSION THREAD

**Discussion by jpmcb** [ghostty-org/ghostty#2705] [Feature Requests, Ideas] [answered]
The ChatGPT desktop app now has a feature where the app can "read" from certain developer programs that have integrated with the desktop app:

- https://help.openai.com/en/articles/10119604-work-with-apps-on-macos
- https://techcrunch.com/2024/11/14/chatgpt-can-now-read-some-of-your-macs-desktop-apps/

Here's a screenshot from it on my desktop:

![Image](https://github.com/user-attachments/assets/cedbf19e-8e4e-4a64-b99a-aa53106a8b74)

---

I'm unsure what's involved with getting this to work with Ghostty. And I'm not seeing developer docs from OpenAI. It may be they need to integrate on their end (in that case, can use this issue to track that integration). I'm also not seeing any sort of integration in VScode: https://github.com/search?q=repo%3Amicrosoft%2Fvscode+chatgpt&type=code

Anyone we know at OpenAI in `ghostty-org` that could assist?

**Comment by markpeek** [comment]
I will note there was pushback ([HN](https://news.ycombinator.com/item?id=40425428)) on the AI feature going into iTerm 3.5.0 and there was a subsequent [release](https://iterm2.com/downloads/stable/iTerm2-3_5_1.changelog) that made it a separate plugin to sandbox the AI code. Likely ghostty should evaluate this approach as well for this type of feature. Closed ghostty plugin related issue: #1358

**Reply by mitchellh** [reply]
This seems to be a different approach where ChatGPT is reading _from_ external applications rather than "AI" (waves hands) being integrated directly into a program. 

If Ghostty can provide the right OS integration where this can work with the ChatGPT app I'm happy to do it. This seems to explicitly require accessibility permissions and the ChatGPT docs make it clear how you can definitely block their app from reading it too. It wouldn't be built-in to Ghostty and would require you already installed ChatGPT too.

I have a feeling the APIs they're using (since they're using AX features) would also enable non-AI applications such as screen readers. It wouldn't be magic there but I think there are some valid use cases this would enable.

So this is likely related to #2351 too.

**Reply by jpmcb** [reply]
> This seems to be a different approach where ChatGPT is reading from external applications rather than "AI" (waves hands) being integrated directly into a program.

+1 - from what I remember of the iTerm AI debacle, people resisted AI inference wrappers being integrated _directly_ into the Terminal (where you could put in your OpenAI / Anthropic token to then have the model directly write commands in the terminal through iTerm's integrations).

Using the AX features seems like a really great approach since it can be opted in / opted out at the system level, works with other desktop apps / screen readers, and leaves the "AI magic" up to those other apps without Ghostty having to integrate.

**Comment by ScriptAlchemist** [comment]
I'd also like this feature. It seems like it just reads around 200 lines from each pane to use somehow. But it's a feature that I would like to use and would have started today if it wouldn't just view the terminal output when I start it inside of ghostty

![image](https://github.com/user-attachments/assets/cc3f2edb-8e56-4cd2-b6e4-1891a0343a11)

**Reply by ScriptAlchemist** [reply]
In using the terminal app with ChatGPT. It's so buggy with my setup on iterm even with their tmux integration and using nvim.

The model rarely has the correct context and it adds things in the scrollback that screw it up in the response. Cause it can't see the whole nvim buffer just the screen view and whatever is above the open vim.

Annoying ai integration growing pains.

If you want it for vim it's likely better to just use a plugin directly at the moment.

**Comment by pluiedev** [comment]
This is a standpoint that's already been reiterated in other places, but Ghostty won't feature any AI integration built-in. Any such support would likely be reserved for a future plugin system.

**Reply by ScriptAlchemist** [reply]
With the creators saying 

"If Ghostty can provide the right OS integration where this can work with the ChatGPT app I'm happy to do it."

In 3 chats above.

**Comment by culstrup** [comment]
Just wanted to bump this w/ a concrete use case: I've found this really useful w/ Terminal when using Claude Code in particular because when I don't totally understand the approach it's taking, I'll ask ChatGPT by the MacOS app to take a look at the Claude Code output as it's working and then explain to me like I'm 5 years old what's going on. 

Being able to do this via Ghostty would be fantastic! Having that real-time tutor (without the copying and pasting) has been critical for me actually learning and not just "vibe coding."

**Reply by jcollie** [reply, marked as answer]
I will reiterate that the current plan for AI features is to not put them in the core. Instead independent developers can add AI support once Ghostty has plugin support (sometime after 1.2 is eventually released).
